File: apps/sepuling/src/test/java/smecalculus/bezmen/storage/SepulkaDaoIT.java
package smecalculus.bezmen.storage;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import smecalculus.bezmen.construction.SepulkaDaoBeans;
import smecalculus.bezmen.core.ServerSideEg;

@ExtendWith(SpringExtension.class)
@ContextConfiguration(classes = SepulkaDaoBeans.class)
@Sql("/schemas/sepulkarium/truncate.sql")
abstract class SepulkaDaoIT {

    @Autowired
    private SepulkaDao sepulkaDao;

    @Test
    void shouldAddOneSepulka() {
        // given
        var expected1 = ServerSideEg.aggregateState().build();
        // and
        var expected2 =
                ServerSideEg.creationState().internalId(expected1.internalId()).build();
        // when
        var actualSaved = sepulkaDao.add(expected1);
        // and
        var actualSelected = sepulkaDao.getBy(expected1.externalId());
        // then
        assertThat(actualSaved)
                .usingRecursiveComparison()
                .isEqualTo(expected1);
        // and
        assertThat(actualSelected).contains(expected2)
                .usingRecursiveComparison();
    }

    @Test
    void shouldViewOneSepulka() {
        // given
        var aggregate = ServerSideEg.aggregateState().build();
        // and
        sepulkaDao.add(aggregate);
        // and
        var expected = ServerSideEg.previewState(aggregate).build();
        // when
        var actual = sepulkaDao.getBy(aggregate.internalId());
        // then
        assertThat(actual).contains(expected);
    }

    @Test
    void shouldUpdateOneSepulka() {
        // given
        var aggregate = ServerSideEg.aggregateState().build();
        // and
        sepulkaDao.add(aggregate);
        // and
        var updatedAt = aggregate.updatedAt().plusSeconds(1);
        // and
        var touch = ServerSideEg.touchState(aggregate)
                .updatedAt(updatedAt)
                .build();
        // when
        var updatedCount = sepulkaDao.updateBy(touch, aggregate.internalId());
        // then
        assertThat(updatedCount).isEqualTo(1);
    }
}

File: apps/sepuling-kotlin/src/test/kotlin/smecalculus/bezmen/storage/SepulkaDaoIT.kt
package smecalculus.bezmen.storage

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.jdbc.Sql
import org.springframework.test.context.junit.jupiter.SpringExtension
import smecalculus.bezmen.construction.SepulkaDaoBeans
import smecalculus.bezmen.core.StateDmEg

@DirtiesContext
@ExtendWith(SpringExtension::class)
@ContextConfiguration(classes = [SepulkaDaoBeans.Anyone::class])
@Sql("/schemas/sepulkarium/truncate.sql")
abstract class SepulkaDaoIT {
    @Autowired
    private lateinit var sepulkaDao: SepulkaDao

    @Test
    fun shouldAddOneSepulka() {
        // given
        val expected1 = StateDmEg.aggregateRoot().build()
        // and
        val expected2 = StateDmEg.existence().internalId(expected1.internalId).build()
        // when
        val actualSaved = sepulkaDao.add(expected1)
        // and
        val actualSelected = sepulkaDao.getBy(expected1.externalId)
        // then
        assertThat(actualSaved).usingRecursiveComparison().isEqualTo(expected1)
        // and
        assertThat(actualSelected).usingRecursiveComparison().isEqualTo(expected2)
    }

    @Test
    fun shouldViewOneSepulka() {
        // given
        val aggregate = StateDmEg.aggregateRoot().build()
        // and
        sepulkaDao.add(aggregate)
        // and
        val expected = StateDmEg.preview(aggregate).build()
        // when
        val actual = sepulkaDao.getBy(aggregate.internalId)
        // then
        assertThat(actual).isEqualTo(expected)
    }

    @Test
    fun shouldUpdateOneSepulka() {
        // given
        val aggregate = StateDmEg.aggregateRoot().build()
        // and
        sepulkaDao.add(aggregate)
        // and
        val updatedAt = aggregate.updatedAt.plusSeconds(1)
        // and
        val touch = StateDmEg.touch(aggregate).updatedAt(updatedAt).build()
        // when
        sepulkaDao.updateBy(touch, aggregate.internalId)
        // then
        // no exception
    }
}
